core-bundle/src/Twig/Interop/ContextFactory.php

```diff
@@ -0,0 +1,175 @@
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Twig\Interop;
+
+use Contao\Template;
+
+/**
+ * @experimental
+ */
+final class ContextFactory
+{
+    /**
+     * Create a Twig template context from a @see Template object.
+     */
+    public function fromContaoTemplate(Template $template): array
+    {
+        $context = $template->getData();
+
+        array_walk_recursive(
+            $context,
+            function (&$value, $key): void {
+                if ($value instanceof \Closure) {
+                    $value = $this->getCallableWrapper($value, (string) $key);
+                }
+            }
+        );
+
+        if (!isset($context['Template'])) {
+            $context['Template'] = $template;
+        }
+
+        return $context;
+    }
+
+    /**
+     * Create a Twig template context from an arbitrary object. This will also
+     * make protected methods/properties/constants accessible.
+     */
+    public function fromClass(object $object): array
+    {
+        $class = new \ReflectionClass($object);
+
+        $context = iterator_to_array($this->getAllMembers($object));
+
+        foreach ($class->getReflectionConstants() as $constant) {
+            if ($constant->isPrivate()) {
+                continue;
+            }
+
+            $context[$constant->getName()] = $constant->getValue();
+        }
+
+        foreach ($class->getStaticProperties() as $property => $value) {
+            if ($class->getProperty($property)->isPrivate()) {
+                continue;
+            }
+
+            $context[$property] = $value;
+        }
+
+        foreach ($class->getMethods() as $method) {
+            if ($method->isPrivate()) {
+                continue;
+            }
+
+            $name = $method->getName();
+
+            if (0 === strpos($name, '__')) {
+                continue;
+            }
+
+            $context[$name] = $this->getCallableWrapper($method->getClosure($object), $name);
+        }
+
+        if (!isset($context['data'])) {
+            $context['data'] = $object;
+        }
+
+        return $context;
+    }
+
+    /**
+     * Find all members including those that were dynamically set ($this->foo = 'bar').
+     */
+    private function getAllMembers(object $object): \Generator
+    {
+        // See https://externals.io/message/105697#105697
+        // Backwards compatibility with PHP < 7.4
+        $mangledObjectVars = \function_exists('get_mangled_object_vars') ?
+            get_mangled_object_vars($object) : (array) $object;
+
+        foreach ($mangledObjectVars as $key => $value) {
+            if (preg_match('/^\x0(\S+)\x0(\S+)$/', $key, $matches)) {
+                // Protected member
+                if ('*' === $matches[1]) {
+                    yield $matches[2] => $value;
+                }
+
+                continue;
+            }
+
+            // Public or dynamic member
+            yield $key => $value;
```

Comment:
```suggestion
            if (strncmp($key, "\0*\0", 3) === 0) {
                // Protected member
                $key = substr($key, 3);
            }

            if ("\0" !== $key[0]) {
                yield $key => $value;
            }
```
Nitpick ☺️

Reply:
Right, after the private members are gone, this looks simpler now :smile: